### PR TITLE
feat(#major): Level finance subgraph; add Level Finance subgraph

### PR DESCRIPTION
**Pull request by shashwatS22**
Deployment links:
1. Arbitrum: https://thegraph.com/hosted-service/subgraph/shashwats22/level-finance-arbitrum
2. BSC: https://thegraph.com/hosted-service/subgraph/shashwats22/level-finance-testing

### CODE REVIEW

File: subgraphs/gmx/src/mappings/GlpManager.ts
Comment (dmelotik) [inline subgraphs/gmx/src/mappings/GlpManager.ts:1]: These `gmx` changes are out of scope of this PR

Comment (shashwatS22) [reply subgraphs/gmx/src/mappings/GlpManager.ts:1]: I think they got pushed by mistake, my bad, I'll remove them and push the code.

**Comment by dmelotik** [comment]
@shashwatS22 a couple notes before I start my review:

- can you fix the lint/build issues
- Update the versions in `deployment.json`
- Mark the PR as `Ready for Review` once that is done

Thanks!

**Comment by shashwatS22** [comment]
> @shashwatS22 a couple notes before I start my review:
> 
> * can you fix the lint/build issues
> * Update the versions in `deployment.json`
> * Mark the PR as `Ready for Review` once that is done
> 
> Thanks!

Got the QA results today, will add the QA fix and will address these notes with them, thanks!

File: subgraphs/level-finance/package.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "level-finance",
+  "license": "UNLICENSED",
+  "scripts": {
+    "prepare:constants": "mustache protocols/${npm_config_protocol}/config/deployments/${npm_config_id}/configurations.json src/common/constants.mustache > src/common/constants.ts"
+  },
+  "dependencies": {
+    "@graphprotocol/graph-cli": "0.40.0",
```
Comment (dmelotik) [inline subgraphs/level-finance/package.json:8]: nit: https://www.npmjs.com/package/@graphprotocol/graph-cli at 0.54.0 now

File: subgraphs/level-finance/protocols/level-finance/config/templates/level.finance.template.yaml [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        - name: ChainLinkContract
+          file: ./abis/Prices/ChainLink.json
+        # Uniswap Contracts
+        - name: UniswapRouter
+          file: ./abis/Prices/Uniswap/Router.json
+        - name: UniswapFactory
+          file: ./abis/Prices/Uniswap/Factory.json
+        - name: UniswapPair
+          file: ./abis/Prices/Uniswap/Pair.json
+
```
Comment (dmelotik) [inline subgraphs/level-finance/protocols/level-finance/config/templates/level.finance.template.yaml:67]: nit: remove newline

File: subgraphs/level-finance/src/sdk/protocols/perpfutures/account.ts
Comment (dmelotik) [inline subgraphs/level-finance/src/sdk/protocols/perpfutures/account.ts:1]: Use the latest version of the SDK `1.1.6`, you should be able to just copy/paste the new code since the params and outputs should be the same

File: subgraphs/level-finance/src/common/utils.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+export function getOutputTokenPrice(pool: Pool): BigDecimal {
+  const tranchesAddresses = pool.pool._tranches;
+  let pricesSum = constants.BIGDECIMAL_ZERO;
+
+  for (let i = 0; i < tranchesAddresses!.length; i++) {
+    const tranche = getOrCreateTranche(tranchesAddresses![i]);
+    pricesSum = pricesSum.plus(
+      tranche!.tvl.div(
+        bigIntToBigDecimal(tranche!.totalSupply, constants.DEFAULT_DECIMALS)
```
Comment (dmelotik) [inline subgraphs/level-finance/src/common/utils.ts:159]: Do all tranche tokens have 18 decimals?

File: subgraphs/level-finance/src/common/utils.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  const lpTokenContract = LpTokenContract.bind(trancheAddress);
+  const totalSupply = readValue(
+    lpTokenContract.try_totalSupply(),
+    constants.BIGINT_ZERO
+  );
+  return totalSupply;
+}
+
+export function getOutputTokenPrice(pool: Pool): BigDecimal {
+  const tranchesAddresses = pool.pool._tranches;
```
Comment (dmelotik) [inline subgraphs/level-finance/src/common/utils.ts:152]: Should we check this exists? If not, that would be a major error, more of a sanity check

File: subgraphs/level-finance/src/common/utils.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  for (let i = 0; i < tranchesAddresses!.length; i++) {
+    const tranche = getOrCreateTranche(tranchesAddresses![i]);
+    totalSupply = totalSupply.plus(tranche.totalSupply);
+  }
+  if (tranchesAddresses!.length <= 0) return constants.BIGINT_ZERO;
+  return totalSupply;
+}
+
+export function roundToWholeNumber(n: BigDecimal): BigDecimal {
+  return n.truncate(0);
```
Comment (dmelotik) [inline subgraphs/level-finance/src/common/utils.ts:182]: I think truncate is essentially a floor function. It only rounds down. Is it okay that it won't round up?

File: subgraphs/level-finance/src/common/utils.ts (hunk 4 above)
Comment (shashwatS22) [reply subgraphs/level-finance/src/common/utils.ts:152]: Added a check, thank you for mentioning this!

File: subgraphs/level-finance/src/common/utils.ts (hunk 3 above)
Comment (shashwatS22) [reply subgraphs/level-finance/src/common/utils.ts:159]: Yes, but if you want we can add the tranche as a token and use the token.decimals for it.

File: subgraphs/level-finance/src/common/utils.ts (hunk 5 above)
Comment (shashwatS22) [reply subgraphs/level-finance/src/common/utils.ts:182]: Yes, I used the MasterChef code from the balancer's code, because they are using essentially the same contract and this roundToWholeNumber function was from that code, so it should be fine.

File: subgraphs/level-finance/package.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "level-finance",
+  "license": "UNLICENSED",
+  "scripts": {
+    "prepare:constants": "mustache protocols/${npm_config_protocol}/config/deployments/${npm_config_id}/configurations.json src/common/constants.mustache > src/common/constants.ts"
```
Comment (dmelotik) [inline subgraphs/level-finance/package.json:5]: this can be removed. It is baked into `messari-subgraph-cli`

Comment (dmelotik) [reply subgraphs/level-finance/package.json]: @shashwatS22 I apologize, this is needed

**Review by dmelotik — APPROVED** [review]
Nice work @shashwatS22 ! Once the build is fixed, I will merge and deploy!

File: subgraphs/level-finance/package.json (hunk 6 above)
Comment (shashwatS22) [reply subgraphs/level-finance/package.json:5]: No worries, I reverted it.

**Comment by shashwatS22** [comment]
> Nice work @shashwatS22 ! Once the build is fixed, I will merge and deploy!

Thanks! PR is ready to merge.
